StyleCop.Analyzers/StyleCop.Analyzers.Test/LayoutRules/SA1518UnitTests.cs

```diff
@@ -0,0 +1,152 @@
+        /// <summary>
+        /// Verifies that a single carriage return / linefeed at the end of the file will not produce a warning.
+        /// </summary>
+        /// <returns></returns>
+        [TestMethod, TestCategory("MaintainabilityRules/SA1518")]
+        public async Task TestWithSingleCarriageReturnLineFeedAtEndOfFile()
+        {
+            var testCode = BaseCode + "\r\n";
+
+            await this.VerifyCSharpDiagnosticAsync(testCode, EmptyDiagnosticResults, CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Verifies that a no carriage return / linefeed at the end of the file will not produce a warning.
+        /// </summary>
+        /// <returns></returns>
+        [TestMethod, TestCategory("MaintainabilityRules/SA1518")]
+        public async Task TestWithoutCarriageReturnLineFeedAtEndOfFile()
+        {
+            await this.VerifyCSharpDiagnosticAsync(BaseCode, EmptyDiagnosticResults, CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Verifies that a single carriage return / linefeed at the end of the file will not produce a warning.
+        /// </summary>
+        /// <returns></returns>
+        [TestMethod, TestCategory("MaintainabilityRules/SA1518")]
+        public async Task TestFileEndsWithSpaces()
+        {
+            var testCode = BaseCode + "\r\n          ";
+
+            var expected = new[]
+            {
+                new DiagnosticResult
+                {
+                    Id = DiagnosticId,
+                    Message = "Code must not contain blank lines at end of file",
+                    Severity = DiagnosticSeverity.Warning,
+                    Locations = new[] { new DiagnosticResultLocation("Test0.cs", 9, 1) }
+                }
+            };
+
+            await this.VerifyCSharpDiagnosticAsync(testCode, expected, CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Verifies that the code fix provider will strip trailing blank lines.
+        /// </summary>
+        /// <remarks>The CRLF after the last curly bracket will not be stripped!</remarks>
+        /// <returns></returns>
+        [TestMethod, TestCategory("MaintainabilityRules/SA1518")]
+        public async Task TestCodeFixProviderStripsTrailingBlankLines()
+        {
+            var testCode = BaseCode + "\r\n\r\n";
+            var fixedTestCode = BaseCode + "\r\n";
+
+            await this.VerifyCSharpFixAsync(testCode, fixedTestCode);
+        }
+
+        /// <summary>
+        /// Verifies that the code fix provider will strip trailing blank lines that include whitespace.
+        /// </summary>
+        /// <remarks>The CRLF after the last curly bracket will not be stripped!</remarks>
+        /// <returns></returns>
+        [TestMethod, TestCategory("MaintainabilityRules/SA1518")]
+        public async Task TestCodeFixProviderStripsTrailingBlankLinesIncludingWhitespace()
+        {
+            var testCode = BaseCode + "\r\n   \r\n   \r\n";
```

Comment:
:bulb: I'd like to see some additional tests using other trivia, such as comments or preprocessor directives, just to make sure we don't accidentally remove something the user wanted to keep.